DataFormats/Detectors/ZDC/include/DataFormatsZDC/RecEvent.h

```diff
@@ -60,15 +61,68 @@ struct RecEvent {
     mRecBC.back().addTDC();
   }
   // Add event information
-  inline void addInfo(uint8_t ch, uint16_t info)
+  inline void addInfo(uint16_t info)
   {
-    if (ch >= NChannels) {
-      ch = 0x1f;
+#ifdef O2_ZDC_DEBUG
+    printf("addInfo info=%u 0x%04x\n", info, info);
+#endif
+    mInfo.emplace_back(info);
+    mRecBC.back().addInfo();
+  }
+
+  inline void addInfo(uint8_t ch, uint16_t code)
+  {
+    if (ch >= NChannels && ch != 0x1f) {
+      LOGF(ERROR, "Adding info (0x%x) for not existent channel %u", code, ch);
+      return;
     }
-    info = (info & 0x07ff) | ch << 11;
+    uint16_t info = (code & 0x03ff) | ((ch & 0x1f) << 10);
+#ifdef O2_ZDC_DEBUG
+    printf("addInfo ch=%u code=%u info=%u 0x%04x\n", ch, code, info, info);
+#endif
     mInfo.emplace_back(info);
     mRecBC.back().addInfo();
   }
+
+  void addInfo(const std::array<bool, NChannels>& vec, const uint16_t code)
+  {
+    int cnt = 0;
+    for (uint8_t ich = 0; ich < NChannels; ich++) {
```

Comment:
you are already looping once over all channels, why don't you fill in the local `std::array<int, NChannels>` indices of cnt non-empty channels, and then in the further loops, scan only those.